tests/epyccel/test_epyccel_pow.py

```diff
@@ -0,0 +1,110 @@
+import pytest
+from numpy.random import rand, randint, uniform
+from numpy import isclose
+
+from pyccel.decorators import types
+from pyccel.epyccel import epyccel
+from conftest import *
+
+def test_pow_int_int(language):
+    @types(int, int)
+    def f_call(x, y):
+        return x ** y
+
+    f = epyccel(f_call, language=language)
+    x = randint(10)
+    y = randint(10)
+
+    assert f(x, y) == f_call(x, y)
+    # negative base
+    assert f(-x, y) == f_call(-x, y)
+
+    assert isinstance(f(x, y), type(f_call(x, y)))
+
+def test_pow_real_real(language):
+    @types('real', 'real')
+    def pow_r_r(x, y):
+        return x ** y
+
+    f = epyccel(pow_r_r, language=language)
+    x = rand()
+    y = rand()
+
+    assert(isclose(f(x, y), pow_r_r(x, y), rtol=1e-15, atol=1e-15))
+
+    assert isinstance(f(x, y), type(pow_r_r(x, y)))
+
+def test_pow_real_int(language):
+    @types('real', 'int')
+    def pow_r_i(x, y):
+        return x ** y
+
+    f = epyccel(pow_r_i, language=language)
+    x = uniform()
+    y = randint(5)
+
+    assert(isclose(f(x, y), pow_r_i(x, y), rtol=1e-15, atol=1e-15))
+    assert isinstance(f(x, y), type(pow_r_i(x, y)))
+
+def test_pow_int_real(language):
+    @types('int', 'real')
+    def pow_i_r(x, y):
+        return x ** y
+
+    f = epyccel(pow_i_r, language=language)
+    x = randint(40)
+    y = uniform()
+
+    assert(isclose(f(x, y), pow_i_r(x, y), rtol=1e-15, atol=1e-15))
+    assert isinstance(f(x, y), type(pow_i_r(x, y)))
+
+def test_pow_special_cases(language):
+    @types('real', 'real')
+    def pow_sp(x, y):
+        return x ** y
+
+    f = epyccel(pow_sp, language=language)
+    e = uniform()
+    assert(isclose(f(0.0, e), pow_sp(0.0, e), rtol=1e-15, atol=1e-15))
+    assert(isclose(f(0.0, e), pow_sp(0.0, e), rtol=1e-15, atol=1e-15))
+
+# ---------------------------- Complex numbers ----------------------------- #
+
+def test_pow_c_c(language):
+    @types('complex', 'complex')
+    def pow_c_c(x, y):
+        return x ** y
+
+    f = epyccel(pow_c_c, language=language)
+    b = complex(rand(), rand())
+    e = complex(rand(), rand())
+    assert(isclose(f(b, e), pow_c_c(b, e), rtol=1e-15, atol=1e-15))
+    assert(isclose(f(-b, e), pow_c_c(-b, e), rtol=1e-15, atol=1e-15))
+    assert(isclose(f(b, -e), pow_c_c(b, -e), rtol=1e-15, atol=1e-15))
+    assert(isclose(f(-b, -e), pow_c_c(-b, -e), rtol=1e-15, atol=1e-15))
+
+def test_pow_c_i(language):
+    @types('complex', 'int')
+    def pow_c_i(x, y):
+        return x ** y
+
+    f = epyccel(pow_c_i, language=language)
+    b = complex(rand(), rand())
+    e = randint(10)
+    assert(isclose(f(b, e), pow_c_i(b, e), rtol=1e-15, atol=1e-15))
+    assert(isclose(f(-b, e), pow_c_i(-b, e), rtol=1e-15, atol=1e-15))
+    assert(isclose(f(b, -e), pow_c_i(b, -e), rtol=1e-15, atol=1e-15))
+    assert(isclose(f(-b, -e), pow_c_i(-b, -e), rtol=1e-15, atol=1e-15))
+
+def test_pow_c_r(language):
```

Comment:
Could you add a `test_pow_r_c` to check complex exponents without a complex base

Reply:
the test fails. we should cast base to complex.
I tried to use PythonComplex but it doesn't work.
any suggestion?
or should I cast it manually?

Reply:
As discussed multiple times before it is better to avoid manual casting wherever possible.
The solution depends on why it is not working

Reply:
Are you sure you wrote the test correctly? It passes on my computer

Reply:
yes, the test is correct. Fortran passed it but c no.
I just pushed the new test case.

Reply:
Your tests are passing on travis

Reply:
the worst thing to have is this situation.
the test fails from time to time.
I am now testing this special cases 
cpow(-(0.8848429441339932+0.7027477315056795j), 7)

Reply:
This is true. Have you modified the test at all since it failed? Do you remember what the error message said?

Reply:
No, the worst. there is a false result.
```language = 'c'

    def test_pow_c_i(language):
        @types('complex', 'int')
        def pow_c_i(x, y):
            return x ** y
    
        f = epyccel(pow_c_i, language=language)
        b = complex(rand(), rand())
        e = randint(10)
        assert(isclose(f(b, e), pow_c_i(b, e), rtol=1e-15, atol=1e-15))
        assert(isclose(f(-b, e), pow_c_i(-b, e), rtol=1e-15, atol=1e-15))
>       assert(isclose(f(b, -e), pow_c_i(b, -e), rtol=1e-15, atol=1e-15))
E       assert False
E        +  where False = isclose((3094888.768776493-1342386.774745199j), (3094888.7687764983-1342386.774745202j), rtol=1e-15, atol=1e-15)
E        +    where (3094888.768776493-1342386.774745199j) = <built-in function pow_c_i>((0.18802209340404008+0.00855573871939952j), -9)
E        +    and   (3094888.7687764983-1342386.774745202j) = <function test_pow_c_i.<locals>.pow_c_i at 0x7f989e895b80>((0.18802209340404008+0.00855573871939952j), -9)

test_epyccel_pow.py:96: AssertionErro```
the error I mentioned yesterday was when I tried to use casting with PythonComplex.
I thought the false result was because we didn't use casting.

Reply:
If you think the fix is to convert the base to a complex then you need to run:
`git cherry-pick a3fe3ea20f484fefbad158992915a8f80886d3dc`
in order to fix the problem you had before with `PythonComplex`. But I don't think that this is necessary to solve the problem as there are currently tests passing which do not do the cast.

>cpow(-(0.8848429441339932+0.7027477315056795j), 7)

This is not a `test_pow_r_c` or a `test_pow_i_c`. Which is failing?

Reply:
Ah yes, that looks like a precision error. You are probably right that converting to complex might fix it.
Luckily the error message gives you the input values so you can test this failing case:
`(0.18802209340404008+0.00855573871939952j), -9`

Reply:
Try increasing `rtol`  to `1e-14`...

Reply:
> Try increasing `rtol` to `1e-14`...

This seems to be sufficient, at least for the example above

Reply:
@EmilyBourne  I don't think so, c convert to the correct type.
check this [implicit conversion](https://en.cppreference.com/w/c/language/conversion)
the missing casting will just raise some warnings.
rtol 1e-14 works great.

Reply:
> the missing casting will just raise some warnings.

If it raises warnings then it would be good to get rid of those. I don't seem to have warnings on my computer, but it has been known to keep quiet about casting